lib/cancan/ability.rb

```diff
@@ -369,9 +366,30 @@ def relevant_rules(action, subject)
         rule.relevant? action, subject
       end
       relevant.reverse!.uniq!
+      optimize_order! relevant
       relevant
     end
 
+    # Optimizes the order of the rules, so that rules with the :all subject are evaluated first.
+    def optimize_order!(rules)
```

Comment:
It won't. This code puts `[:all]` rules before other rules within consecutive groups of `base_behavior == true` rules. We cannot reorder `true` and `false` clause relative to each other because they are not disjuntive (the order matters, `cannot` takes priority over a previously defined `can`).